crates/miden-lib/asm/account_components/rpo_falcon_512_procedure_acl.masm

```diff
@@ -16,61 +16,115 @@ const.NUM_AUTH_TRIGGER_PROCS_SLOT=1
     push.0
-    # => [was_any_called, i, pad(16)]
+    # => [require_acl_auth, i, pad(16)]
 
     # Loop through trigger procedures
     dup.1 neq.0
     while.true
-        # => [was_any_called, i, pad(16)]
+        # => [require_acl_auth, i, pad(16)]
 
         # Get the procedure root from storage
         dup.1 sub.1 push.0.0.0 push.AUTH_TRIGGER_PROCS_MAP_SLOT
-        # => [AUTH_TRIGGER_PROCS_MAP_SLOT, [0, 0, 0, i-1], was_any_called, i, pad(16)]
+        # => [AUTH_TRIGGER_PROCS_MAP_SLOT, [0, 0, 0, i-1], require_acl_auth, i, pad(16)]
 
         exec.account::get_map_item
-        # => [AUTH_TRIGGER_PROC_ROOT, was_any_called, i, pad(16)]
+        # => [AUTH_TRIGGER_PROC_ROOT, require_acl_auth, i, pad(16)]
 
         exec.account::was_procedure_called
-        # => [was_called, was_any_called, i, pad(16)]
+        # => [was_called, require_acl_auth, i, pad(16)]
 
-        # Update was_any_called
+        # Update require_acl_auth
         or
-        # => [was_any_called', i, pad(16)]
+        # => [require_acl_auth', i, pad(16)]
 
         swap sub.1 swap
-        # => [was_any_called', i-1, pad(16)]
+        # => [require_acl_auth', i-1, pad(16)]
 
         # Check if we should continue looping
         dup.1 neq.0
-        # => [should_continue, was_any_called', i-1, pad(16)]
+        # => [should_continue, require_acl_auth', i-1, pad(16)]
     end
+    # => [require_acl_auth, i-1, pad(16)]
+
+    swap drop
+    # => [require_acl_auth, pad(16)]
+
+    # ------ Check if output notes were created ------
+
+    exec.tx::get_output_notes_commitment
+    # => [OUTPUT_NOTES_COMMITMENT, require_acl_auth, pad(16)]
+
+    padw eqw not movdn.8 dropw dropw
+    # => [were_output_notes_created, require_acl_auth, pad(16)]
 
-    # => [was_any_called, i-1, pad(16)]
+    loc_load.0 not
+    # => [!allow_unauthorized_output_notes, were_output_notes_created, require_acl_auth, pad(16)]
 
-    # If any trigger procedure was called, perform signature verification
+    and
+    # => [require_output_note_auth, require_acl_auth, pad(16)]
+
+    # ------ Check if input notes were consumed ------
+
+    exec.tx::get_input_notes_commitment
+    # => [INPUT_NOTES_COMMITMENT, require_output_note_auth, require_acl_auth, pad(16)]
+
+    padw eqw not movdn.8 dropw dropw
+    # => [were_input_notes_consumed, require_output_note_auth, require_acl_auth, pad(16)]
+
+    loc_load.1 not
+    # => [!allow_unauthorized_input_notes, require_output_note_auth, require_acl_auth, pad(16)]
+
+    and
+    # => [require_input_note_auth, require_output_note_auth, require_acl_auth, pad(16)]
+
+    or
+    # => [require_note_auth, require_acl_auth, pad(16)]
+
+    # Combine note-based auth requirement with procedure-based auth requirement
+    or
+    # => [require_auth, pad(16)]
+
+    # If authentication is required (either note-based or procedure-based), perform signature verification
     if.true
         exec.::miden::contracts::auth::basic::auth__tx_rpo_falcon512
```

Comment:
Rather than calling `auth__tx_rpo_falcon512` (which is meant to be invoked via `call` instruction), it may be better to have some helper functions which are shared between that and this auth component.

Reply:
Agreed, though it may make sense to do this after https://github.com/0xMiden/miden-base/pull/1624 is merged, since it touches a lot of the falcon code.

Reply:
Agreed. Maybe in #1624 we can already split out some things into helper procedures - e.g.,

- `prepare_tx_summary` - puts the 4 words of the tx summary onto the stack.
- `hash_tx_summary` hashes summary.

Reply:
That's pretty much what I ended up with (`create_tx_summary` and `hash_tx_summary`) and they are made public in https://github.com/0xMiden/miden-base/pull/1634, though if it's time-critical that can be easily done as part of a separate PR as well.